src/oidcendpoint/session/database.py

```diff
@@ -0,0 +1,173 @@
+import logging
+from typing import List
+from typing import Union
+
+from . import session_key
+from .grant import Grant
+from .info import ClientSessionInfo
+from .info import SessionInfo
+from .info import UserSessionInfo
+
+logger = logging.getLogger(__name__)
+
+
+class Database(object):
+    def __init__(self, storage=None):
+        if storage is None:
+            self._db = {}
+        else:
+            self._db = storage
+
+    @staticmethod
+    def _eval_path(path: List[str]):
+        uid = path[0]
+        client_id = None
+        grant_id = None
+        if len(path) > 1:
+            client_id = path[1]
+            if len(path) > 2:
+                grant_id = path[2]
+
+        return uid, client_id, grant_id
+
+    def set(self, path: List[str], value: Union[SessionInfo, Grant]):
+        """
+
+        :param path: a list of identifiers
+        :param value: Class instance to be stored
+        """
+        # Try loading the key, that's a good place to put a debugger to
+        #  import pdb; pdb.set_trace()
+        uid, client_id, grant_id = self._eval_path(path)
+
+        _userinfo = self._db.get(uid)
+        if _userinfo:
+            if client_id:
+                if client_id in _userinfo['subordinate']:
+                    _cid_key = session_key(uid, client_id)
+                    _cid_info = self._db[session_key(uid, client_id)]
+                    # if _cid_info.is_revoked():
+                    #     raise Revoked("Session is revoked")
+                    if _cid_info:
+                        if grant_id:
+                            _gid_key = session_key(uid, client_id, grant_id)
+                            if grant_id in _cid_info['subordinate']:
+                                _gid_info = self._db[_gid_key]
+                                if not _gid_info:
+                                    self._db[_cid_key] = _cid_info.add_subordinate(grant_id)
```

Comment:
Isn't the `grant_id` already a `_cid_info subordinate`. It was checked in https://github.com/IdentityPython/oidcendpoint/pull/104/files#diff-93bc9f688c9bce53870f2c48de99c5e6b094668e243a710ce6d72abad9708a79R54

Reply:
Not clear which line/-s you're referring to.
Remember there are 2 places subordinates appear:
1) as a identifier in the list of subordinates to an object. (for example _cid_info['subordinate'])
2) by value in the database (for example self._db[_gid_key])

So, in order to add a subordinate you have to 
1) Add the identifier to the 'subordinate' parameter and
2) Add the information connected to the subordinate in the database